src/goose/synopsis/text_editor.py

```diff
@@ -0,0 +1,160 @@
+        self._log_file_operation(path, content, language)
+        return f"Successfully wrote to {path}"
+
+    def _patch_file(self, path: str, before: str, after: str) -> str:
+        """Patch the file by replacing 'before' with 'after'."""
+        patho = system.to_patho(path)
+
+        if not patho.exists():
+            raise ValueError(f"You can't patch {path} - it does not exist yet")
+        if not system.is_active(path):
+            raise ValueError(f"You must view {path} using read_file before you patch it")
+
+        content = patho.read_text()
+
+        if content.count(before) != 1:
+            raise ValueError("The 'before' content must appear exactly once in the file.")
+
+        self._save_file_history(patho)
+        content = content.replace(before, after)
+        system.remember_file(path)
+        patho.write_text(content)
+
+        self._log_file_operation(path, f"{before} -> {after}", get_language(path))
+        return "Successfully replaced before with after."
+
+    def _save_file_history(self, patho: Path) -> None:
+        """Save the current content of the file to history for undo functionality."""
+        content = patho.read_text() if patho.exists() else ""
+        self._file_history[str(patho)] = content
+
+    def _undo_edit(self, path: str, **kwargs: dict) -> str:
+        """Undo the last edit made to a file."""
+        patho = system.to_patho(path)
+
+        if not patho.exists() or str(patho) not in self._file_history:
+            raise ValueError(f"No edit history available to undo changes on {path}.")
+
+        previous_content = self._file_history.pop(str(patho))
+        patho.write_text(previous_content)
+        system.remember_file(path)
+
+        self._log_file_operation(path, "Undo edit", get_language(path))
+        return f"Successfully undid the last edit on {path}"
+
+    def _view_file_or_directory(self, path: str, view_range: Optional[list[int]] = None, **kwargs: dict) -> str:
+        """View the content of a file or directory."""
+        patho = system.to_patho(path)
+
+        if patho.is_file():
+            return self._view_file(patho, view_range)
+        elif patho.is_dir():
+            return self._view_directory(patho)
+        else:
+            raise ValueError(f"The path {path} does not exist.")
+
+    def _view_file(self, patho: Path, view_range: Optional[list[int]]) -> str:
+        if not patho.exists():
+            raise ValueError(f"The file {patho} does not exist.")
+
+        with open(patho, "r") as f:
+            content = f.readlines()
+
+        if view_range:
+            start_line, end_line = view_range
+            if start_line < 1 or end_line < start_line:
+                raise ValueError("Invalid view range.")
+            content = content[start_line - 1 : (end_line if end_line != -1 else len(content))]
+
+        system.remember_file(str(patho))
+        self._log_file_operation(str(patho), "".join(content), get_language(str(patho)))
```

Comment:
nit: imo, this is too verbose to show every read - it is consistent, but i'd consider only showing this on writes?